### ISSUE TITLE
add support for header parameters

### ISSUE THREAD

**Issue by stakach** [spider-gazelle/action-controller#66]

**Comment by tristanholl** [comment]
I am not sure if it is intentional to use `open_api_param` instead of the `ann_converter`, but this line: https://github.com/spider-gazelle/action-controller/blob/b5d3c139afd9b8325037caf66e9462736af31835/src/action-controller/router/builder.cr#L390 prevents the use of an annotation with the parameter `in: :header`

The annotation:
```Crystal
@[AC::Param::Info(description: "Idempotency key to ensure unique processing", in: :header )]
idempotency_key : UUID
```

still results in
```JSON
"parameters": [
    {
        "name": "idempotency_key",
        "in": "query",
        "description": "Idempotency key to ensure unique processing",
        "required": true,
        "schema": {
            "type": "string",
            "format": "uuid"
        }
    }
]
```

Changing the above mentioned line to 

```Crystal
{% open_api_param[:in] = (ann_converter && ann_converter[:in])] || :query %}
``` 

allows to move route parameters to the header in the OpenAPI spec and results in a correctly generated spec:

```JSON
"parameters": [
    {
        "name": "idempotency_key",
        "in": "header",
        "description": "Idempotency key to ensure unique processing",
        "required": true,
        "schema": {
            "type": "string",
            "format": "uuid"
        }
    }
]
```

I have not noticed any side-effects, but this part is not tested well enough.

**Comment by stakach** [comment]
At the time this was put together we desperately needed good API documentation and headers as parameters wasn't a high priority.
I have higher priorities the next few weeks but usually have some time around new years as things quiet down.

Definitely open to pull requests though so here is a rough draft of what needs to happen if you have time:

The code is quite complex as it looks for all the parameters and where they come from (route or query currently) at compile time.
This involves going through all the before, after around filters etc and collecting any param usage along the way and recording where the param is located.
This is what `open_api_route[:params]["param_name"][:in]` is storing

That macro information is used to extract the required information from the various locations when a request is made to a route and pass the parsed information to your apps handler functions.
The extraction code for each route is generated at compile time for maximum speed at runtime. Hence the complexity, although this could probably be simplified once we have [macro defs](https://github.com/crystal-lang/crystal/issues/8835)

So for your example code above this would work better (as you'd need to be defining what the header name is)

```crystal
# extract this param from the `X-My-UUID` header
@[AC::Param::Info(description: "Idempotency key to ensure unique processing", header: "X-My-UUID" )]
idempotency_key : UUID
```

Then to generate the docs correctly, something like this (need to capture the header name information)

```crystal
{% if (ann_converter && ann_converter[:header]) %}
  {% open_api_param[:in] = :header %}
  {% open_api_param[:header] = ann_converter[:header] %}
{% else %}
  {% open_api_param[:in] = open_api_param[:in] || :query %}
{% end %}
```

then the [extraction code](https://github.com/spider-gazelle/action-controller/blob/b5d3c139afd9b8325037caf66e9462736af31835/src/action-controller/router/builder.cr#L448-L483) would need updating
After the check for `body_argument` you could then do something like
```
{% elsif open_api_param[:header] %}
   # extract the param_value from the header and process it
```

Then finally the [documentation generation](https://github.com/spider-gazelle/action-controller/blob/b5d3c139afd9b8325037caf66e9462736af31835/src/action-controller/open_api.cr#L348) code would need to be updated

There is a [comment here](https://github.com/spider-gazelle/action-controller/blob/b5d3c139afd9b8325037caf66e9462736af31835/src/action-controller/open_api/parameter.cr#L6) that indicates that the OpenAPI spec requires slightly different formatting for header params too

**Comment by stakach** [comment]
It shouldn't be `nil` when there is a route param present
https://github.com/spider-gazelle/action-controller/blob/master/src/action-controller/router/builder.cr#L312-L324

i.e. id in this route `@[AC::Route::GET("/route/:id/component")]`

**Comment by stakach** [comment]
fixed in b60101affda37dcb908531d16552f510340c5538
